inspector/src/components/graph/canvasGraphService.ts

```diff
@@ -416,9 +459,131 @@ export class CanvasGraphService {
      */
     private _removeEventListeners(canvas: HTMLCanvasElement) {
         canvas.removeEventListener("wheel", this._handleZoom);
+        canvas.removeEventListener("mousemove", this._handleDataHover);
         canvas.removeEventListener("mousedown", this._handlePanStart);
+        canvas.removeEventListener("mouseleave", this._handleStopHover);
         canvas.ownerDocument.removeEventListener("mouseup", this._handlePanStop);
     }
+    
+    /**
+     * Handles what to do when we are hovering over the canvas and not panning.
+     * 
+     * @param event A reference to the event to be handled. 
+     */
+    private _handleDataHover = (event: MouseEvent) => {
+        if (this._panPosition) {
+            // we don't want to do anything if we are in the middle of panning
+            return;
+        }
+
+        this._hoverPosition = event.clientX;
+
+        // then draw the tooltip.
+        this._drawTooltip(this._hoverPosition, this._drawableArea);
+    }
+
+    /**
+     * Handles what to do when we stop hovering over the canvas.
+     */
+    private _handleStopHover = () => {
+        this._hoverPosition = null;
+    }
+
+    /**
+     * Draws the tooltip given the area it is allowed to draw in and the current pixel position.
+     * 
+     * @param pixel the position of the mouse cursor in pixels. 
+     * @param drawableArea  the available area we can draw in.
+     */
+    private _drawTooltip(pixel: number | null, drawableArea: IGraphDrawableArea) {
+        const {_ctx : ctx} = this;
+
+        if (pixel === null || !ctx || !ctx.canvas) {            
+            return;
+        }
+
+        // first convert the mouse position in pixels to a timestamp.
+        const {left: start, right: end} = ctx.canvas.getBoundingClientRect();
+        const inferredTimestamp = this._getNumberFromPixel(pixel, this._globalTimeMinMax, start, end);
+        
+        const results: IPerfTooltip[] = [];
+
+        // get the closest timestamps to the target timestamp, and store the appropriate meta object.
+        this.datasets.forEach((dataset: IPerfDataset) => {

```

Comment:
Same as we discussed last time, This should try to be garbage collection friendly

Reply:
Would 2 arrays work in this case instead of a tooltip array we have 2 arrays of strings?

Reply:
Even if you keep only one array you could reuse it from move to move instead of recreating it on every events

Reply:
Ahh okay so its the array creation that is the main issue here rather than the object creation in the loop? So doing something similar to what I do with the ticks array (instance variable then .length = 0 in the event) should be fine then?

Reply:
you might want to keep the allocated array around with the max potential size with the objects in and keep a local length to know the meaningful part of the data. Then in here you could update the length and data content for the meaningful part. This would prevent allocating and such for the array as well as the GC involved by the creation of the new objects.

Reply:
Latest commit pre-creates the objects in the constructor, and we loop over and define the meaningful part in the event. We simply update the objects that have been pre-allocated now!